Supabase guide (#203)
sequinstream/sequin: docs/guides/supabase.mdx

<Steps>
  <Step title="Create a replication slot">
    In the SQL editor, run the following SQL command to create a replication slot:

    ```sql
    select pg_create_logical_replication_slot('sequin_slot', 'pgoutput');
    ```

    This creates a replication slot named `sequin_slot` that Sequin will use to capture changes from your database.
  </Step>
  <Step title="Create a publication">
    Next, you'll create a publication to indicate which tables will publish changes to the replication slot.

    In the SQL editor, run the following SQL command to create a publication:

    ```sql
    create publication sequin_pub for all tables;
    ```

    If you want to publish changes from specific tables, you can specify them in the publication.

    ```sql
    create publication sequin_pub for table table1, table2, table3;
    ```
  </Step>
  <Step title="Enter the replication slot details in Sequin">
    Back in the Sequin Console, enter the name of the replication slot (e.g. `sequin_slot`) and publication (e.g. `sequin_pub`) you just created. Then, name your database and click **Create Database**.
  </Step>
</Steps>

> <Check>Sequin will now capture changes from your Supabase database.</Check>

Reviewer: ```suggestion
<Check>Sequin will now capture rows and changes from your Supabase database.</Check>
```

## Moving Supabase webhooks to Sequin

> Sequin adds some helpful features to Supabase webhooks:

Reviewer: ```suggestion
Sequin has some helpful features that Supabase webhooks lack:

```

* Exactly-once processing guarantees ensuring every event is processed once and only once.
* Automatic retries and backoff for failed events.

> * Filtering of events with no plpgsql.

Reviewer: ```suggestion
* Filtering of events with no PL/pgSQL.
```

* Backfill support to sync historical data.
* Helpful observability and debugging tools.

> You can easily move your existing Supabase webhooks to Sequin by creating Sequin HTTP push consumers. Here is a quick example of how to do this using a hypothetical `orders` table:

Reviewer: HTTP push consumers  -> Link to docs section

```suggestion
You can easily move your existing Supabase webhooks to Sequin by creating Sequin HTTP push consumers. Here is an example of how to do this using a hypothetical `orders` table:
```

> ### Supabase webhook

Reviewer: This section header confused me

Reviewer: Ah, looks like you're contrasting the two - I think that could be way clearer

> Everytime an new order is created, a postgres trigger fires a webhook to send the data to the `orders` endpoint:

Reviewer: ```suggestion
Everytime a new order is created, a Supabase Postgres trigger fires a webhook to send the data to the `orders` endpoint:
```

Reviewer: Still confused, why is this sentence ending with a `:`?

> Here is the configuration for the webhook in the Supabase:

Reviewer: grammar

<Tabs>
  <Tab title="SQL">
  ```sql
    create trigger new_order
    after insert on public.orders
    for each row
    execute function supabase_functions.http_request ('https://webhook.site/4f1af4fe-ce43-4530-9d75-f0f2abc086f6', 'POST', '{"Content-type":"application/json"}', '{}', '1000')
    ```
  </Tab>
  <Tab title="Supabase Console">
    <Frame>
        <img src="../images/guides/supabase/supabase-webhook.png" alt="Supabase webhook" />
    </Frame>
  </Tab>
</Tabs>

> When this trigger fires, the data is sent to the `orders` endpoint with the following payload:

Reviewer: use active voice, not passive

```json
{
  "type": "INSERT",
  "table": "orders",
  "record": {
    "id": 4,
    "user_id": 4,
    "order_date": "2024-09-03T18:32:48.108729+00:00",
    "total_amount": 154.97
  },
  "schema": "public",
  "old_record": null
}
```

### Sequin HTTP push consumer

To create this exact webhook in Sequin, you'll create an HTTP push consumer on the `orders` table with no filters. Here's how to do it:

<Steps>
  <Step title="Create a new push consumer">
    In the Sequin Console, navigate to the Consumers page, click **Create Consumer**.

    In the setup wizard, select **Change stream** as the stream type and then **HTTP Push** as the consumer type.
  </Step>
  <Step title="Select your database and table">
    Select your Supabase database and the `public.orders` table.
  </Step>
  <Step title="Set filters and operations">

>     Don't configure any filters and set the consumer to capture just the `INSERT` operations:

Reviewer: ```suggestion
    Don't configure any filters and set the consumer to capture just `INSERT` operations:
```

    <Frame>
        <img src="../images/guides/supabase/sequin-consumer-filters.png" alt="Sequin consumer settings" />
    </Frame>
  </Step>
  <Step title="Add your webhook URL">
    Finally, give your consumer a name, set a timeout, and add your webhook URL:
    <Frame>
        <img src="../images/guides/supabase/sequin-consumer-endpoint.png" alt="Sequin consumer settings" />
    </Frame>

>     <Note>Sequin gives you more control over webhook delivery and retries including a higher request timeout, automatic retries, and the ability to configure max ack pending (e.g. the consumer will stop sending new events to the endpoint if a certain number of events are not acknowledged). For even more control over event consumption, you can use Sequin's HTTP pull consumer.</Note>

Reviewer: Not sure the sell to the pull consumer is needed

  </Step>
</Steps>

Now, when you create a new order in Supabase, Sequin will capture the change and send it to your endpoint:

```json
{
  "record": {
    "id": 5,
    "order_date": "2024-09-03T19:41:17Z",
    "total_amount": "154.97",
    "user_id": 5
  },
  "metadata": {
    "consumer": {
      "id": "107adb3d-76b4-40ba-8e9d-587f9871ab5c",
      "name": "new_order"
    },
    "table_name": "orders",
    "table_schema": "public",
    "commit_timestamp": "2024-09-03T19:41:17.650384Z"
  },
  "action": "insert",
  "changes": null
}
```

<Note>
  Sequin's data payload is slightly different to Supabase's webhook payload:
  * Sequin's `metadata` field includes additional metadata about the consumer and the table.
  * Sequin captures the operation type (e.g. `insert`, `update`, `delete`) in the `action` field not the `type` field.

>   * Sequin captures changes for `update` operations in the `changes` field not the `old_record` field.

Reviewer: I assume `changes` is also different than `old_record`? Do you need to explain `changes` here?

</Note>

## Triggering a Supabase edge function with Sequin

> Sequin can also trigger Supabase edge functions by calling the function's endpoint URL. This is an alternative to triggering Supabase edge functions with a Supabase webhooks, but with some nice guarantees and observability. Notably, Sequin can retry failed events.

Reviewer: ```suggestion
Sequin can trigger Supabase edge functions by calling the function's endpoint URL. This is an alternative to triggering Supabase edge functions with Supabase webhooks. Sequin improves on the Supabase workflow by offering more features, like retries for failed events.
```

<Steps>
  <Step title="Update the edge function to process Sequin's payload">

>     Update the edge function to process Sequin's event payload. Note that while Sequin's record payload is the same, the changes, actions, and metadata are different:

Reviewer: Interesting that this section is about _migrating_ vs adopting. I think that's wise, but could be clearer that this is what this section (and whole guide?) is about

Reviewer: ```suggestion
    Update the edge function to process Sequin's event payload. Note that while Sequin's `record` payload is the same, the keys `changes`, `actions`, and `metadata` are different:
```

    <CodeGroup>
        ```json Supabase
        {
        "type": "INSERT",
        "table": "orders",
        "record": {
            "id": 4,
            "user_id": 4,
            "order_date": "2024-09-03T18:32:48.108729+00:00",
            "total_amount": 154.97
        },
        "schema": "public",
        "old_record": null
        }
        ```

        ```json Sequin
        {
        "record": {
            "id": 5,
            "order_date": "2024-09-03T19:41:17Z",
            "total_amount": "154.97",
            "user_id": 5
        },
        "metadata": {
            "consumer": {
            "id": "107adb3d-76b4-40ba-8e9d-587f9871ab5c",
            "name": "new_order"
            },
            "table_name": "orders",
            "table_schema": "public",
            "commit_timestamp": "2024-09-03T19:41:17.650384Z"
        },
        "action": "insert",
        "changes": null
        }
        ```
    </CodeGroup>
  </Step>
  <Step title="Deploy the edge function and retrieve the URL">
    Using the Supabase CLI, deploy your edge function:

    ```bash

>     supabase functions deploy {your-function-name} --no-verify-jwt

Reviewer: Is there any security on this endpoint at all? Any way to add security? That is a requirement.

    ```

    <Note>As with Supabase webhooks, you'll need to deploy your edge function with the `--no-verify-jwt` flag to allow Sequin to call the function.</Note>

    Once deployed, you can retrieve the URL of the edge function by logging into the Supabase Dashboard and navigating to the Edge Functions page:

    <Frame>
        <img src="../images/guides/supabase/supabase-function-url.png" alt="Supabase function URL" />
    </Frame>

  </Step>
  <Step title="Create a Sequin push consumer">
    With your edge function URL in hand, you 'll create a Sequin push consumer and set the endpoint to the edge function URL:

    <Frame>
        <img src="../images/guides/supabase/sequin-supabase-edge-function.png" alt="Sequin consumer settings" />
    </Frame>
  </Step>
</Steps>

<Check>Now, when you make a change in your Supabase database that meets the consumer's filter criteria, Sequin will capture the change and send it to the edge function.</Check>

## Next steps

Sequin is a great complement to Supabase. This guide showed you how to connect Supabase to Sequin and start moving your webhooks and edge function triggers to Sequin. From here, you may want to:

> * Refine your webhooks / triggers by using more specific filters. Perhaps you can move logic out of your edge function and into your consumer filters.

Reviewer: Nice

> * Use Sequin's HTTP pull consumer to gain more control over event consumption.

Reviewer: Link
